streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java

```
@@ -16,350 +16,173 @@
     }
 
     /**
      * @throws TaskAssignmentException if method fails to encode the data
      */
     public ByteBuffer encode() {
-        final ByteBuffer buf;
-
-        switch (usedVersion) {
-            case 1:
-                buf = encodeVersionOne();
-                break;
-            case 2:
-                buf = encodeVersionTwo();
-                break;
-            case 3:
-                buf = encodeVersionThree();
-                break;
-            case 4:
-                buf = encodeVersionFour();
-                break;
-            default:
-                throw new IllegalStateException("Unknown metadata version: " + usedVersion
-                    + "; latest supported version: " + LATEST_SUPPORTED_VERSION);
-        }
-
-        buf.rewind();
-        return buf;
-    }
-
-    private ByteBuffer encodeVersionOne() {
-        final ByteBuffer buf = ByteBuffer.allocate(getVersionOneByteLength());
-
-        buf.putInt(1); // version
-        encodeClientUUID(buf);
-        encodeTasks(buf, prevTasks);
-        encodeTasks(buf, standbyTasks);
-
-        return buf;
-    }
-
-    private int getVersionOneByteLength() {
-        return 4 + // version
-               16 + // client ID
-               4 + prevTasks.size() * 8 + // length + prev tasks
-               4 + standbyTasks.size() * 8; // length + standby tasks
-    }
-
-    protected void encodeClientUUID(final ByteBuffer buf) {
-        buf.putLong(processId.getMostSignificantBits());
-        buf.putLong(processId.getLeastSignificantBits());
-    }
-
-    protected void encodeTasks(final ByteBuffer buf,
-                               final Collection<TaskId> taskIds) {
-        buf.putInt(taskIds.size());
-        for (final TaskId id : taskIds) {
-            id.writeTo(buf);
-        }
-    }
-
-    protected byte[] prepareUserEndPoint() {
-        if (userEndPoint == null) {
-            return new byte[0];
+        if (data.version() > LATEST_SUPPORTED_VERSION) {
```

Am I right in thinking that this can't happen, given our validation in the constructor?

---

Yeah from looking at the validation code, I'm not sure we'd hit this point.  But if we did wouldn't we want to throw an exception as we do in the validation section? Although I think we may use a number higher greater than latest supported version for version probing.

---

Yeah, I'm not sure... Well, this is what the prior code did, so I guess we can just leave it alone.

---

Given we've done `validateVersions` at constructor shouldn't this never happen? In that case we should also throw right?

---

Yeah, I think it shouldn't happen. I'll swap in the exception and re-run all the tests.